src/components/recs/rec-tile.tsx

```diff
@@ -1,34 +1,164 @@
+            {
+                d[numElementsToConsider][targetSum] = true;
+            }
+            // If we get an exact solution, stop everything
+            if (targetSum == targ && d[numElementsToConsider][targ])
+            {
+                foundExactSolution = true;
+                bestSolutionTarget = targ;
+                bestSolutionNumElements = numElementsToConsider;
+                break;
+            }
+            if (d[numElementsToConsider][targetSum] && thisError < bestSolutionError)
+            {
+                bestSolutionError = thisError;
+                bestSolutionTarget = targetSum;
+                bestSolutionNumElements = numElementsToConsider;
+            }
+        }
+        if (foundExactSolution) break;
+    }
+    let sumRemaining = bestSolutionTarget;
+    let valuesPicked: number[] = [];
+    let startNumElements = Math.min(matrixXDim, bestSolutionNumElements);
+    for (let numElementsToConsider=startNumElements; numElementsToConsider > 0; numElementsToConsider--)
+    {
+        if (!d[numElementsToConsider][sumRemaining])
+        {
+            throw new Error(`Subset sum failed: matrix pos ${numElementsToConsider} ${sumRemaining} should be true`)
+        }
+        // Find the boundary where making this sum isn't possible any more
+        while (d[numElementsToConsider-1][sumRemaining])
+        {
+            numElementsToConsider--;
+        }
+        // and get that value
+        const thisVal = vals[numElementsToConsider-1];
+        valuesPicked.push(thisVal);
+        sumRemaining -= thisVal;
+        if (sumRemaining <= 0)
+            break;
+    }
+
+    return valuesPicked;
+}
+
+/**
+ * Split team data into two, one for each side of the map image.
+ * @returns array of two elements: the first containing the indexes of the teams that should be on the left,
+ *          and the second containing the indexes of teams that should be on the right.
+ */
+function splitTeams(metadata: IRecordedGame): number[][]
+{
+  // Fast track the case of two teams
+  const numTeams = metadata.teams.length;
+  if (numTeams == 2)
+  {
+    return [[0], [1]];
+  }
+  const target = Math.round(metadata.gameNumPlayers/2);
+  const teamSizes = metadata.teams.map((team) => { return team.length });
+  const optimalSizes = subsetSumClosest(teamSizes, target);
+  const setOne: number[] = [];
+  const setTwo: number[] = [];
+  for (let teamIndex=0; teamIndex<teamSizes.length; teamIndex++)
+  {
+    const teamSize = teamSizes[teamIndex];
+    if (optimalSizes.indexOf(teamSize) > -1)
+    {
+      setOne.push(teamIndex);
+      optimalSizes.splice(optimalSizes.indexOf(teamSize), 1);
+    }
+    else
+    {
+      setTwo.push(teamIndex);
+    }
+  }
+  // For consistency, put the set with fewer teams on the left
+  if (setOne.length < setTwo.length)
+  {
+    return [setOne, setTwo];
+  }
+  return [setTwo, setOne];
+}
```

Comment:
Looks good - can you move all of this logic out of the react component? Want to keep business logic out of the front end code as much as possible. Maybe add a teams.ts somewhere in /server